Enable typechecking of shadow memory primitives

esteffin opened the pull request:
This PR enables typechecking of shadow memory primitives.

The primitives are checked for correct number of arguments, correct types and also enforce other invariants such as having a constant string for name, a bit-vector type of size `< 8` and non-`void` pointers for accessing it.

It also adds regressions and unit tests for it.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add regression tests for SM typecheck.

Commit:
Adding unit tests for SM typecheck.

Commit:
Fixed SM failing regressions tests.

Fixed SM failing regression tests due to better error messaging on
typecheck.

peterschrammel commented on src/ansi-c/c_typecheck_shadow_memory_builtin.cpp:
```diff
@@ -0,0 +1,250 @@
+  }
+
+  // The second argument type must be either a boolean or a bitvector of size
+  // less than the size of a char.
```
less or equal

esteffin replied:
```suggestion
  // less or equal to the size of a char.
```

peterschrammel commented on regression/cbmc-shadow-memory/intrinsics_warnings1/intrinsics_warn.c:
```diff
@@ -0,0 +1,32 @@
+  assert(__CPROVER_get_field(&s.j, "uninitialized") == 1);
+
+  return 0;
+}
```
newline at the end (several occurrences)

esteffin replied:
Done

peterschrammel commented on src/ansi-c/c_typecheck_shadow_memory_builtin.cpp:
```diff
@@ -0,0 +1,250 @@
+  if(expr.arguments().size() != 2)
+  {
+    std::ostringstream error_message;
+    error_message << identifier << " takes exactly 2 argument, but "
```
```suggestion
    error_message << identifier << " takes exactly 2 arguments, but "
```

peterschrammel commented on src/ansi-c/c_typecheck_shadow_memory_builtin.cpp:
```diff
@@ -0,0 +1,250 @@
+  if(expr.arguments().size() != 2)
+  {
+    std::ostringstream error_message;
+    error_message << identifier << " takes exactly 2 argument, but "
```
```suggestion
    error_message << identifier << " takes exactly 2 arguments, but "
```

peterschrammel commented on src/ansi-c/c_typecheck_shadow_memory_builtin.cpp:
```diff
@@ -0,0 +1,250 @@
+  if(expr.arguments().size() != 3)
+  {
+    std::ostringstream error_message;
+    error_message << identifier << " takes exactly 3 argument, but "
```
```suggestion
    error_message << identifier << " takes exactly 3 arguments, but "
```

peterschrammel commented on src/ansi-c/c_typecheck_shadow_memory_builtin.cpp:
```diff
@@ -0,0 +1,250 @@
+    // Can't set the shadow memory value with a non-boolean and non-bitvector
+    // value.
+    std::ostringstream error_message;
+    error_message << identifier << " argument 3 must be a bit-field, but ("
```
❓ bit-field?

esteffin replied:
```suggestion
    error_message << identifier << " argument 3 must be a byte-sized integer, but ("
```

Commit:
Adding typecheck code for shadow memory builtins.

Adding typecheck code for shadow memory builtins so they are
type-checked (more thoroughly than with the C typechecker) and
correctly converted to function calls.
